### DISCUSSION TITLE
Why is generic_bfs_edges not in __all__?

### DISCUSSION THREAD

**Discussion by mathiasbockwoldt** [networkx/networkx#5907] [Q&A] [answered]
I need to perform a bfs on a directed graph but ignoring the direction. I would like to have the behaviour of `bfs_edges` and not `edge_bfs` (seriously confusing naming...). After a lot of struggling, I found `generic_bfs_edges` in the source, but not in the documentation. With the generic `neighbors` parameter, this seems like the right function for my purpose.

Is there any reason that `generic_bfs_edges` is not in `__all__`? It is fully documented and all, but since it is not in `__all__` it does not show up the docs and importing it is more complicated and feels a bit like using an "underscore" function.

**Comment by dschult** [comment, marked as answer]
The function `generic_bfs_edges` is old enough that it doesn't have the underscore prefix, but it is in fact a private function. That said, as you mention it is fully documented and all.  The function signature may change in a future release -- indeed we have recently discussed combining the input arguments `neighbors` and `sorted_neighbors` (I might have the name wrong). The other functions use it and it isn't going anywhere.  You can access it via:

```python
nx.breadth_first_search.generic_bfs_edges(...)
```

All that said, you could (more easily?) use `bfs_edges(G.to_undirected())`. That probably how we intended you to get what you want.

**Reply by mathiasbockwoldt** [reply]
Thank you very much for the detailed answer!

Could you say something about the performance of G.to_undirected()? I'm currently working on graphs with around 250 000 nodes and even more edges. I thought about keeping a permanent undirected copy in memory anyway, but I have to be a bit careful with memory as well (I have multiple graphs open at any time). 

One thing I really like about the generic version is the `neighbors` parameter. Sometimes, I want to create a subgraph by spreading out from a target node until I find nodes with certain properties. Using the `neighbors` parameter, I can elegantly achieve this by passing a suitable function.

Well, if the interface might change, I will probably write my own generic bfs function (based on yours) for local use.

By the way, I used `from networkx.algorithms.traversal.breadth_first_search import generic_bfs_edges`, but your usage is, of course, much more concise.

**Reply by dschult** [reply]
The keyword `as_view` Input is a low memory version of to_undirected:   `bfs_edges(G.to_undirected(as_view=True))`
It is a view on the original graph rather than a copy. That likely takes longer computation time, but not too much more and it definitely saves memory.  I would suggest that for your case.

**Reply by mathiasbockwoldt** [reply]
Thank you! I will see how the undirected view performs.
